schematic/store/synapse.py

```diff
@@ -51,9 +51,98 @@
 from schematic.exceptions import MissingConfigValueError, AccessCredentialsError
 
 from schematic import CONFIG
+from schematic.utils.general import entity_type_checking
 
 logger = logging.getLogger("Synapse storage")
 
+class ManifestDownload(object):
+    def _download_manifest_to_folder(self, syn, manifest_id):
+        """
+        try downloading a manifest to local cache or a given folder
+        manifest
+        Args:
+            syn: an object of type synapseclient.
+            manifest_id: id of a manifest  
+        Return: 
+            manifest_data: A new Synapse Entity object of the appropriate type
+        """
+        if CONFIG["synapse"]["manifest_folder"]:
+            manifest_data = syn.get(
+                    manifest_id,
+                    downloadLocation=CONFIG["synapse"]["manifest_folder"],
+                    ifcollision="overwrite.local",
+                )
+        else:
+            manifest_data = syn.get(
+                        manifest_id,
+                    )
+        return manifest_data 
+
+    def _entity_type_checking(self, syn, manifest_id):
+        """
+        check the entity type of the id that needs to be downloaded
+        Args:
+            syn: an object of type synapseclient.
+            manifest_id: id of a manifest
+        Return: 
+             if the entity type is wrong, raise an error
+        """
+        # check the type of entity
+        entity_type = entity_type_checking(syn, manifest_id)
+        if entity_type  != "file":
+            logger.error(f'You are using entity type: {entity_type}. Please try using a file')
+
+    @staticmethod
+    def download_manifest(self, syn, manifest_id: str, newManifestName="", manifest_df=None):
+        """
+        Donwload a manifest based on a given manifest id. 
+        Args:
+            syn: an object of type synapseclient.
+            manifest_id: id of a manifest
+            newManifestName(optional): new name of a manifest that gets downloaded.
+            manifest_df(optional): a dataframe containing name and id of manifests in a given asset view
+        Return: 
+            manifest_data: synapse entity file object
+        """
+
+        # enables retrying if user does not have access to uncensored manifest
+        # pass synID to synapseclient.Synapse.get() method to download (and overwrite) file to a location
+        manifest_data = ""
+
+        # check entity type
+        self._entity_type_checking(syn, manifest_id)
+
+        # download a manifest
+        try:
+            manifest_data = self._download_manifest_to_folder(syn, manifest_id)
+        except(SynapseUnmetAccessRestrictions, SynapseAuthenticationError):
+            if not manifest_df.empty:
```

Comment:
Can you explain this?  Are you downloading the manifest that is censored?

Reply:
Thanks for the question Tom. This is I guess an edge case that I noticed that the old code is dealing with and I tried to preserve the old behavior. For example, for the dataset here: https://www.synapse.org/#!Synapse:syn27600053, we have two manifests: one is censored and one is uncensored. The uncensored one (see [here](https://www.synapse.org/#!Synapse:syn29862078)) has access restriction, so the old code would try to download the censored manifest.

Reply:
Thanks, makes sense. Do you expect only to have 1 or 2 manifest in a folder?

Reply:
Yes. For a given dataset, there should only be 0-2 manifest.